node/src/types/sync_leap.rs

```diff
@@ -290,20 +295,1676 @@ impl FetchItem for SyncLeap {
+                signed_block_header_with_metadata_1,
+                signed_block_header_with_metadata_2,
+                signed_block_header_with_metadata_3,
+            ],
+        };
+        let actual_headers: BTreeSet<_> = sync_leap
+            .switch_blocks_headers()
+            .map(|header| header.block_hash())
+            .collect();
+        let expected_headers: BTreeSet<_> = [
+            trusted_block,
+            trusted_ancestor_1,
+            signed_block_1,
+            signed_block_2,
+        ]
+        .iter()
+        .map(|block| *block.hash())
+        .collect();
+        assert_eq!(expected_headers, actual_headers);
+    }
+
+    #[test]
+    fn should_return_highest_block_header_from_trusted_block() {
+        // Chain
+        // 0   1   2   3   4   5   6   7   8   9   10   11
+        // S           S           S           S
+        let switch_blocks = [0, 3, 6, 9];
+
+        let mut rng = TestRng::new();
+
+        let query = 5;
+        let trusted_ancestor_headers = [4, 3];
+        let signed_block_headers = [6, 9, 11];
+        let add_proofs = true;
+        let sync_leap = make_test_sync_leap(
+            &mut rng,
+            &switch_blocks,
+            query,
+            &trusted_ancestor_headers,
+            &signed_block_headers,
+            add_proofs,
+        );
+
+        // `sync_leap` is a well formed SyncLeap structure for the test chain. We can use the blocks
+        // it contains to generate SyncLeap structures as required for the test, because we know the
+        // heights of the blocks in the test chain as well as their sigs.
+        let highest_block = sync_leap
+            .signed_block_headers
+            .last()
+            .unwrap()
+            .block_header
+            .clone();
+        let lowest_blocks: Vec<_> = sync_leap
+            .trusted_ancestor_headers
+            .iter()
+            .take(2)
+            .cloned()
+            .collect();
+        let middle_blocks: Vec<_> = sync_leap
+            .signed_block_headers
+            .iter()
+            .take(2)
+            .cloned()
+            .collect();
+
+        let highest_block_height = highest_block.height();
+        let highest_block_hash = highest_block.block_hash();
+
+        let sync_leap = SyncLeap {
+            trusted_ancestor_only: false,
+            trusted_block_header: highest_block.clone(),
+            trusted_ancestor_headers: lowest_blocks,
+            signed_block_headers: middle_blocks,
+        };
+        assert_eq!(
+            sync_leap
+                .highest_block_header_and_signatures()
+                .0
+                .block_hash(),
+            highest_block.block_hash()
+        );
+        assert_eq!(sync_leap.highest_block_hash(), highest_block_hash);
+        assert_eq!(sync_leap.highest_block_height(), highest_block_height);
+    }
+
+    #[test]
+    fn should_return_highest_block_header_from_trusted_ancestors() {
```

Comment:
I'm not sure we need this test do we?  It's testing for something which is impossible from a valid `SyncLeap` unless I'm misunderstanding the intent?

Reply:
As long as `highest_block_header_and_signatures()` references `self.trusted_block_header` I think there's a merit in having this scenario covered.